src/utils.jl

```diff
@@ -95,31 +95,14 @@ function to_extent(edges::Vector{Edge})
     Extents.Extent(X=x, Y=y)
 end
 
-function to_points(x, ::Type{T} = Float64) where T
-    points = Vector{TuplePoint{T}}(undef, _npoint(x))
-    _to_points!(points, x, 1)
-    return points
-end
+_to_point3(T, p) = (T(GI.x(p)), T(GI.y(p)), T(GI.z(p)))
+_to_point2(T, p) = (T(GI.x(p)), T(GI.y(p)))
 
-_to_points!(points::Vector, x, n) = _to_points!(points, trait(x), x, n)
-function _to_points!(points::Vector, ::FeatureCollectionTrait, fc, n)
-    for f in GI.getfeature(fc)
-        n = _to_points!(points, f, n)
-    end
-end
-_to_points!(points::Vector, ::FeatureTrait, f, n) = _to_points!(points, GI.geometry(f), n)
-function _to_points!(points::Vector, ::AbstractGeometryTrait, fc, n)
-    for f in GI.getgeom(fc)
-        n = _to_points!(points, f, n)
-    end
-end
-function _to_points!(points::Vector, ::Union{AbstractCurveTrait,MultiPointTrait}, geom, n)
-    n = 0
-    for p in GI.getpoint(geom)
-        n += 1
-        points[n] = _tuple_point(p)
-    end
-    return n
+"""
+    to_points(geom::AbstractGeometry, ::Type{T} = Float64) where {T <: Number}
+"""
+function to_points(geom, ::Type{T} = Float64) where {T <: Number}
+    return collect(flatten(Base.Fix1(_is3d(geom) ? _to_point3 : _to_point2, T), GI.PointTrait, geom))
```

Comment:
You probably want the conditional outside the generator... does this compile away to be only called once or is it called every time? 

If its outside we know it does. Its worth the duplication of the line anyway.

Reply:
I think that by definition it's called only once - the if statement is simply deciding which function to pass in, no?  It's not an if on the point but on the geometry.

Reply:
No :)

Its inside a Fix2 so its called for every single point that `flatten` calls `f(geom)` on.

For ArchGDAL that will be painfully slow.